genson/src/test/java/com/owlike/genson/convert/JsonDeserializationTest.java

```diff
@@ -357,6 +357,47 @@ public void testDeserializeEnum() {
     assertEquals("foo", pojo.str);
   }
 
+  @Test
+  public void testDeserializationOptional() {
+    assertFalse(genson.deserialize("{\"@class\":\"java.util.Optional\"}", Optional.class).isPresent());
+    assertFalse(genson.deserialize("{\"@class\":\"java.util.Optional\"}", Optional.class).isPresent());
+    assertEquals(Optional.of("string"),
+                 genson.deserialize("{\"@class\":\"java.util.Optional\",\"value\":\"string\"}", Optional.class));
+    assertEquals(Optional.of(42L),
+                 genson.deserialize("{\"@class\":\"java.util.Optional\",\"value\":42}", Optional.class));
+    assertEquals(Optional.of(42.35D),
+                 genson.deserialize("{\"@class\":\"java.util.Optional\",\"value\":42.35}", Optional.class));
+    assertEquals(Optional.of(true),
+                 genson.deserialize("{\"@class\":\"java.util.Optional\",\"value\":true}", Optional.class));
+    assertEquals(Optional.of(false),
+                 genson.deserialize("{\"@class\":\"java.util.Optional\",\"value\":false}", Optional.class));
+
+    Genson genson = new GensonBuilder().useClassMetadata(true).create();
+    assertEquals(Optional.of(new BeanWithConstructor("Bilbo Baggins", 111, null)),
+                 genson.deserialize("{\"@class\":\"java.util.Optional\",\"value\":{\"@class\":\"com.owlike.genson.convert.JsonDeserializationTest$BeanWithConstructor\",\"age\":111,\"name\":\"Bilbo Baggins\",\"other\":null}}", Optional.class));
+  }
+  
+  @Test
+  public void testDeserializationOptionalInt() {
+    assertFalse(genson.deserialize("{\"@class\":\"java.util.OptionalInt\"}", OptionalInt.class).isPresent());
+    assertEquals(OptionalInt.of(42),
+        genson.deserialize("{\"@class\":\"java.util.OptionalInt\",\"value\":42}", OptionalInt.class));
+  }
+
+  @Test
+  public void testDeserializationOptionalLong() {
+    assertFalse(genson.deserialize("{\"@class\":\"java.util.OptionalLong\"}", OptionalLong.class).isPresent());
+    assertEquals(OptionalLong.of(42L),
+        genson.deserialize("{\"@class\":\"java.util.OptionalLong\",\"value\":42}", OptionalLong.class));
+  }
+
+  @Test
+  public void testDeserializationOptionalDouble() {
+    assertFalse(genson.deserialize("{\"@class\":\"java.util.OptionalDouble\"}", OptionalDouble.class).isPresent());
+    assertEquals(OptionalDouble.of(42D),
+        genson.deserialize("{\"@class\":\"java.util.OptionalDouble\",\"value\":42.0}", OptionalDouble.class));
+  }
+
   public static class BeanWithConstructor {
```

Comment:
I would add `BeanWithOptionals` class as well, containing all 4 optionals as fields, and a test that checks whether deserialization of a bean with optionals as fields works properly.